File: aoc2025/go/challenges/day3.go
package challenges

import (
	"fmt"
	"log"
	"log/slog"
	"strconv"
	"strings"

	"aoc/utilities"

	logger "github.com/Bparsons0904/goLogger"
)

type BatteryPack []int

func Day3() {
	log := logger.New("Day3")
	slog.Info("day3")

	batteryPacks := processDay3File()

	timer := log.Timer("Part 1 Timer")
	part1Result := calculate2PackJoltage(batteryPacks)
	timer()

	timer = log.Timer("Part 2 Timer")
	part2Result := calculate12PackJoltage(batteryPacks)
	timer()

	log.Info("Results", "Part 1", part1Result, "Part 2", part2Result)
}

func calculate12PackJoltage(batteryPacks []BatteryPack) int {
	maxJoltage := 0
	for _, batteryPack := range batteryPacks {
		maxJoltage += batteryPack.getLargest12PackJoltage()
	}

	return maxJoltage
}

func (b BatteryPack) getLargest12PackJoltage() int {
	maxJoltage := make([]int, 0, 12)

	index := 0
	iteration := 0
	for {
		if len(maxJoltage) == 12 {
			break
		}
		maxFoundJoltage, endIndex := iterateBatteryPackCheck(b, index, iteration)
		maxJoltage = append(maxJoltage, maxFoundJoltage)
		index = endIndex + 1
		iteration++
	}

	var string strings.Builder
	for _, joltage := range maxJoltage {
		string.WriteString(strconv.Itoa(joltage))
	}

	result, err := strconv.Atoi(string.String())
	if err != nil {
		log.Fatal(err)
	}

	return result
}

func iterateBatteryPackCheck(battery BatteryPack, startingIndex int, batterySize int) (int, int) {
	maxFoundJoltage := 0
	endIndex := startingIndex
	stop := len(battery) - 11 + batterySize
	for i := startingIndex; i < stop; i++ {
		if battery[i] > maxFoundJoltage {
			maxFoundJoltage = battery[i]
			endIndex = i
		}
	}

	return maxFoundJoltage, endIndex
}

func calculate2PackJoltage(batteryPacks []BatteryPack) int {
	maxJoltage := 0
	for _, batteryPack := range batteryPacks {
		maxJoltage += batteryPack.getLargest2PackJoltage()
	}

	return maxJoltage
}

func (b BatteryPack) getLargest2PackJoltage() int {
	maxJoltageFirst := 0
	maxJoltageSecond := 0
	index := 0

	for i := 0; i < len(b)-1; i++ {
		if b[i] > maxJoltageFirst {
			maxJoltageFirst = b[i]
			index = i
		}
	}

	for i := index + 1; i < len(b); i++ {
		if b[i] > maxJoltageSecond {
			maxJoltageSecond = b[i]
		}
	}

	stringedInt := fmt.Sprintf("%d%d", maxJoltageFirst, maxJoltageSecond)
	maxJoltage, err := strconv.Atoi(stringedInt)
	if err != nil {
		log.Fatal(err)
	}

	return maxJoltage
}

func processDay3File() []BatteryPack {
	file := utilities.ReadFile("day3.part1")
	var batteryPacks []BatteryPack
	for _, file := range file {
		batteryPacks = append(batteryPacks, getBatteryPack(file))
	}

	return batteryPacks
}

func getBatteryPack(batteryPackString string) BatteryPack {
	var batteryPack BatteryPack
	for _, batteryString := range batteryPackString {
		battery, err := strconv.Atoi(string(batteryString))
		if err != nil {
			log.Fatal(err)
		}
		batteryPack = append(batteryPack, battery)
	}

	return batteryPack
}

File: aoc2025/zig/src/days/day3.zig
const std = @import("std");
const file_parser = @import("../file_parser.zig");
const print = std.debug.print;

const BatteryPack = []u32;

pub fn run() !void {
    const allocator = std.heap.page_allocator;
    
    var batteryPacks = try allocator.alloc(BatteryPack, 16);
    var batteryPacks_count: usize = 0;
    
    try processDay3File(allocator, &batteryPacks, &batteryPacks_count);
    
    const part1Result = try calculate12PackJoltage(batteryPacks[0..batteryPacks_count], 2);
    const part2Result = try calculate12PackJoltage(batteryPacks[0..batteryPacks_count], 12);
    
    print("Results Part1: {d}, Part2: {d}\n", .{ part1Result, part2Result });
    
    // Free all allocated memory
    for (batteryPacks[0..batteryPacks_count]) |pack| {
        allocator.free(pack);
    }
    allocator.free(batteryPacks);
}

fn calculate12PackJoltage(batteryPacks: []const BatteryPack, batterySize: usize) !u64 {
    var maxJoltage: u64 = 0;
    for (batteryPacks) |batteryPack| {
        maxJoltage += try getLargestPackJoltage(batteryPack, batterySize);
    }
    return maxJoltage;
}

fn processDay3File(allocator: std.mem.Allocator, batteryPacks: *[]BatteryPack, batteryPacks_count: *usize) !void {
    const file = try file_parser.readFile(allocator, "day3.part1");
    defer allocator.free(file);
    
    for (file) |line| {
        var batteryPack = try allocator.alloc(u32, line.len);
        
        for (line, 0..) |batteryChar, i| {
            const battery = try std.fmt.charToDigit(batteryChar, 10);
            batteryPack[i] = @intCast(battery);
        }
        
        if (batteryPacks_count.* >= batteryPacks.len) {
            const new_packs = try allocator.realloc(batteryPacks.*, batteryPacks.len * 2);
            batteryPacks.* = new_packs;
        }
        batteryPacks.*[batteryPacks_count.*] = batteryPack;
        batteryPacks_count.* += 1;
    }
}

// Extension method for BatteryPack
fn getLargestPackJoltage(batteryPack: BatteryPack, batterySize: usize) !u64 {
    var maxJoltage = try std.heap.page_allocator.alloc(u32, batterySize);
    defer std.heap.page_allocator.free(maxJoltage);
    var maxJoltageCount: usize = 0;
    
    var index: usize = 0;
    var iteration: usize = 0;
    
    while (maxJoltageCount < batterySize) {
        const result = try iterateBatteryPackCheck(batteryPack, index, iteration, batterySize);
        maxJoltage[maxJoltageCount] = result.joltage;
        index = result.endIndex + 1;
        iteration += 1;
        maxJoltageCount += 1;
    }
    
    // Convert array of numbers to a single integer
    var result: u64 = 0;
    for (maxJoltage[0..maxJoltageCount]) |joltage| {
        var buf: [20]u8 = undefined;
        const joltageStr = std.fmt.bufPrint(&buf, "{d}", .{joltage}) catch continue;
        for (joltageStr) |digit| {
            const digitVal = try std.fmt.charToDigit(digit, 10);
            result = result * 10 + @as(u64, @intCast(digitVal));
        }
    }
    
    return result;
}

const JoltageResult = struct {
    joltage: u32,
    endIndex: usize,
};

fn iterateBatteryPackCheck(
    battery: BatteryPack,
    startingIndex: usize,
    currentBatterySize: usize,
    batterySize: usize,
) !JoltageResult {
    var maxFoundJoltage: u32 = 0;
    var endIndex: usize = startingIndex;
    const stop = battery.len + currentBatterySize - (batterySize - 1);
    
    var i = startingIndex;
    while (i < stop and i < battery.len) : (i += 1) {
        if (battery[i] > maxFoundJoltage) {
            maxFoundJoltage = battery[i];
            endIndex = i;
        }
    }
    
    return JoltageResult{
        .joltage = maxFoundJoltage,
        .endIndex = endIndex,
    };
}